x-pack/solutions/security/plugins/security_solution/docs/rfcs/detection_response/customized_rule_alert_telemetry.md

```diff
@@ -0,0 +1,273 @@
+
+---
+
+## Proposed Solution: Adding a field to the rule schema to track specific customized fields
+
+### Approach and technical details
+At its core, this solution is pretty simple: adding a new field to the rule schema that would keep track of which field(s) were customized on the rule object. This would allow us to write the customized field list to the alert object at rule execution the same as we do any rule field, without any runtime calculations that'd slow down rule execution. The field would be updated in the same places `rule_source` is currently updated (basically any CRUD operation on the rule object), and would be either omitted or defaulted to an empty array if a rule's base version didn't exist as there'd be no way to accurately calculate the field list.
+
+This field would be an array and it would contain a list of objects that have . It was considered to have this field be a key/value pair with the key being the field name and the value be the original, unmodified rule field but was rejected. Having the original values would no doubt be helpful in some cases but would also introduce a whole host of new edge cases and potentially double the size of the rule object if most fields were customized. Having the names of the fields customized and whether or not they're functional fields is all we'd need for our use case and some of the previously linked telemetry tickets. Furthermore, the querying of an array within the alerts telemetry cluster would be far more straightforward than querying for keys on an object, as KQL doesn't support the direct querying of object keys and we'd have to rely on wildcard queries or some other syntax. Arrays also allow us to easily chain together multiple clauses in order to be extremely granular with our telemetry analysis. I've listed some example KQL queries below. The addition of the `is_functional` field in the object would be there primarily for filtering purposes. Having a way to easily filter out all alerts from rules with customized functional fields makes for a better, less error prone method for writing queries than having to chain potentially 20+ field names together. It also helps with the extensibility of these queries, as anytime we add a field to the rule object we wouldn't have to update the telemetry and related dashboards to account for the rule, we'd just have to set it properly on rule update.
```

Comment:
Not sure if this is the best place to ask this. The `is_functional` field is in the document itself. What happens if we decide alert on that a field is functional after all? All existing docs that have a customized field will continue the old decision and the data will be incorrect. Are we deciding that fields shouldn't / can't be changed in the future?

Reply:
If we switch the single source of truth for a functional field, the data that didn't use that definition would become stale - at least in terms of the "functional_fields" metadata. There are ways around that as we'd have access to the full array of field names for all rules, so if you really needed to, you could just make your own custom filter with x amount of fields in the query and analyze rules that way.

Either way, best to change the functional fields list as little as possible from when we define it